File: articles/calculus-for-systems-modeling/articles/eulers-method/c/euler_method_audit.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

static double exact_solution(double t, double y0, double k){ return y0 * exp(-k * t); }

int main(void){
  const double y0 = 100.0, k = 0.35, h = 0.1, stop_time = 20.0;
  const int steps = (int)round(stop_time / h);
  double y = y0;
  double multiplier = 1.0 - h * k;
  const char *status = fabs(multiplier) <= 1.0 ? "stable_for_simple_decay" : "unstable_risk";

  printf("step,time,euler_value,exact_value,absolute_error,step_size,stability_multiplier,stability_status,warning\n");
  for(int step=0; step<=steps; step++){
    double t = step * h;
    double exact = exact_solution(t, y0, k);
    printf("%d,%.6f,%.12f,%.12f,%.12f,%.6f,%.12f,%s,Euler estimates depend on time step rate function initial condition stability and accumulated error.\n",
      step, t, y, exact, fabs(y - exact), h, multiplier, status);
    y = y + h * (-k * y);
  }
  return EXIT_SUCCESS;
}

File: articles/calculus-for-systems-modeling/articles/eulers-method/fortran/euler_method_audit.f90
program euler_method_audit
  implicit none
  integer :: step, steps
  real(8) :: y0, k, h, stop_time, y, t, exact, multiplier

  y0 = 100.0d0
  k = 0.35d0
  h = 0.1d0
  stop_time = 20.0d0
  steps = nint(stop_time / h)
  y = y0
  multiplier = 1.0d0 - h * k

  print '(A)', 'step time euler_value exact_value absolute_error step_size stability_multiplier'
  do step = 0, steps
    t = dble(step) * h
    exact = y0 * exp(-k * t)
    print '(I6,6F16.8)', step, t, y, exact, abs(y - exact), h, multiplier
    y = y + h * (-k * y)
  end do
end program
